plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/StandardColumnMappings.java

```diff
@@ -445,16 +447,18 @@ public static ColumnMapping timestampColumnMappingUsingSqlTimestamp(TimestampTyp
         return ColumnMapping.longMapping(
                 timestampType,
                 (resultSet, columnIndex) -> {
-                    Timestamp timestamp = resultSet.getTimestamp(columnIndex);
-                    return toPrestoTimestamp(timestampType, timestamp.toLocalDateTime());
+                    LocalDateTime timestamp = resultSet.getTimestamp(columnIndex).toLocalDateTime();
+                    int roundedNanos = toIntExact(round(timestamp.getNano(), 9 - timestampType.getPrecision()));
+                    LocalDateTime roundedTimestamp = timestamp
+                            .withNano(roundedNanos % toIntExact(NANOSECONDS_PER_SECOND))
+                            .plusSeconds(roundedNanos / NANOSECONDS_PER_SECOND);
+                    return toPrestoTimestamp(timestampType, roundedTimestamp);
                 },
-                timestampWriteFunctionUsingSqlTimestamp(timestampType));
-    }
-
-    @Deprecated
-    public static ColumnMapping timestampColumnMapping()
-    {
-        return timestampColumnMapping(TIMESTAMP_MILLIS);
+                timestampWriteFunctionUsingSqlTimestamp(timestampType),
+                // NOTE: pushdown is disabled because the values stored in remote system might not match the values as
+                // read by Trino due to rounding. This can lead to incorrect results if operations are pushed down to
+                // the remote system.
+                DISABLE_PUSHDOWN);
```

Comment:
I liked the rename of this method to `timestampColumnMappingUsingSqlTimestampWithRounding` to indicate the semantics change. It's safe, since there are no other usages. It doesn't need to be pretty, since it's deprecated and likely to go away at some point.